File: SupportPrefenceDemo/app/src/main/java/edu/cs4730/supportprefencedemo/MainActivity.java
package edu.cs4730.supportprefencedemo;

import androidx.annotation.NonNull;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentResultListener;
import androidx.fragment.app.FragmentTransaction;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import android.os.Bundle;

import edu.cs4730.supportprefencedemo.databinding.ActivityMainBinding;


/**
 * This is a simple example of using the support library PreferenceFragmentCompat.
 * The MainFragment has a button to launch the SettingsFragment which shows the
 * preferences, which are saved automatically to SharedPreferences.  This showing older methods of
 * display summary information.  Supportpreference2demo is a newer methods.
 * <p>
 * It can't use the Arch Navigation component because it doesn't support
 * PreferenceFragmentCompat (as of this writing anyway 9/10/2025 ).  So the code is little gross in main.
 * <p>
 * Big NOT here :
 * Must specify preferenceTheme in theme, it's an item.  see the style.xml file.
 */
public class MainActivity extends AppCompatActivity implements MainFragment.OnFragmentInteractionListener,
    PreferenceFragmentCompat.OnPreferenceStartFragmentCallback {
    FragmentManager fragmentManager;
    ActivityMainBinding binding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        ViewCompat.setOnApplyWindowInsetsListener(binding.container, (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return WindowInsetsCompat.CONSUMED;
        });
        fragmentManager = getSupportFragmentManager();
        //setup the mainFragment to show.
        fragmentManager.beginTransaction().add(binding.container.getId(), new MainFragment()).commit();
    }

    /**
     * This the callback interface for MainFragment.
     */
    @Override
    public void onFragmentInteraction(int which) {

        //now change to the SecondFragment, pressing the back button should go to main fragment.
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        //Change to the correct fragment for preferences
        if (which == 1)
            transaction.replace(binding.container.getId(), new PreferenceupdateFragment());
        else
            transaction.replace(binding.container.getId(), new myPreferenceFragment());
        // and add the transaction to the back stack so the user can navigate back
        transaction.addToBackStack(null);

        // Commit the transaction
        transaction.commit();
    }

    /**
     * this is "required" as of version 1.1.0 of the androidx.preference.
     * It used when the user taps on preference that calls a preference fragment.
     */
    @Override
    public boolean onPreferenceStartFragment(@NonNull PreferenceFragmentCompat caller, Preference pref) {
        // Instantiate the new Fragment
        final Bundle args = pref.getExtras();
        final Fragment fragment = fragmentManager.getFragmentFactory().instantiate(
            getClassLoader(), pref.getFragment());
        fragment.setArguments(args);

        //Now replace the existing Fragment with the new Fragment, the listener is doing?? idk.
        fragmentManager.beginTransaction().replace(binding.container.getId(), fragment).addToBackStack(null).commit();
        fragmentManager.setFragmentResultListener("reequestKey", this, new FragmentResultListener() {
            @Override
            public void onFragmentResult(@NonNull String requestKey, @NonNull Bundle result) {
                if (requestKey.equals("requestKey"))
                    fragmentManager.setFragmentResult(requestKey, result);
            }
        });
        return true;

    }
}

File: SupportPrefenceDemo_kt/app/src/main/java/edu/cs4730/supportprefencedemo_kt/MainActivity.kt
package edu.cs4730.supportprefencedemo_kt

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.FragmentManager
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import edu.cs4730.supportprefencedemo_kt.databinding.ActivityMainBinding

/**
 * this a redux of the preferenceDemo, but using the androidx.preferencefragmentCompat (prefupdateFrag and prefNewFrag)
 * and v14 prefrencefragment (myPreferenceActivity and prefFrag  since not a support fragment.
 * No there is no need version compatibility, except a couple of v14 specific preferences cant be used before API 14.
 * for all the new preferences xml.
 * <p>
 * maybe useful as well http://developer.android.com/guide/topics/ui/settings.html
 * <p>
 * Big NOT here :
 * Must specify preferenceTheme in theme, it's an item.  see the style.xml file.
 */

class MainActivity : AppCompatActivity(), MainFragment.OnFragmentInteractionListener,
    PreferenceFragmentCompat.OnPreferenceStartFragmentCallback {
    lateinit var fragmentManager: FragmentManager
    lateinit var binding: ActivityMainBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        ViewCompat.setOnApplyWindowInsetsListener(binding.container) { v: View, insets: WindowInsetsCompat ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            WindowInsetsCompat.CONSUMED
        }

        fragmentManager = supportFragmentManager
        //setup the mainFragment to show.
        fragmentManager.beginTransaction().add(binding.container.id, MainFragment()).commit()
    }

    /**
     * This the callback interface for MainFragment.
     */
    override fun onFragmentInteraction(which: Int) {

        //now change to the SecondFragment, pressing the back button should go to main fragment.
        val transaction = fragmentManager.beginTransaction()
        //Change to the correct fragment for preferences
        if (which == 1) transaction.replace(binding.container.id, PreferenceupdateFragment())
        else transaction.replace(binding.container.id, myPreferenceFragment())
        // and add the transaction to the back stack so the user can navigate back
        transaction.addToBackStack(null)
        // Commit the transaction
        transaction.commit()
    }

    /**
     * this is "required" as of version 1.1.0 of the androidx.preference.
     * It used when the user taps on preference that calls a preference fragment.
     */
    override fun onPreferenceStartFragment(
        caller: PreferenceFragmentCompat, pref: Preference
    ): Boolean {
        // Instantiate the new Fragment
        val args = pref.extras
        val fragment = fragmentManager.fragmentFactory.instantiate(classLoader, pref.fragment!!)
        fragment.arguments = args

        //Now replace the existing Fragment with the new Fragment, the listener is doing?? idk.
        fragmentManager.beginTransaction().replace(binding.container.id, fragment)
            .addToBackStack(null).commit()
        fragmentManager.setFragmentResultListener(
            "reequestKey", this
        ) { requestKey, result ->
            if (requestKey == "requestKey") fragmentManager.setFragmentResult(requestKey, result)
        }
        return true
    }
}
